### PR TITLE
Deno: init at 1.0.0

### LINKED ISSUE #78461: Deno
**Project description**
Deno - A secure JavaScript and TypeScript runtime

**Metadata**

* homepage URL: https://deno.land/
* source URL: https://github.com/denoland/deno
* license: mit
* platforms: unix, linux, darwin

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Motivation for this change

Adding `deno` for it's `1.0.0` release

Resolves: #78461
Tracking issue for `1.0.0` release: https://github.com/denoland/deno/issues/2473
PR for `1.0.0` release: https://github.com/denoland/deno/pull/5273
Release post: <https://deno.land/v1>

~**Version is currently `1.0.0-rc3` as that's the latest release**. I will force-push once 1.0.0 is out (ETA 13th May)~

Any and all suggestions and criticisms welcome, most of my nix packaging experience is on my own machine so I'd like to know if I'm straying from best/good practices.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS (x86_64)
   - [x] macOS
   - [X] NixOS or other (aarch64 ARM)
    - Build passed, Build + Check ended SIGKILL 9 (Raspbian RPI4)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

**Concerns**

* ~Currently using built zip~ Now building from source, not rebuilding v8 though
  * ~Probably preferable to build from source releases but had issues with `rusty_v8`~ Solved
  * Building from source + also building v8 from source is probably not preferred as v8 takes ages to build (maybe have it as optional?)
* ~The release has a `.d.ts` definition~
  * ~I haven't done much actual work with `deno` yet so IDK if I should install this somewhere~
  * ~The [deno_install](https://github.com/denoland/deno_install) script doesn't seem to touch it so it should be fine~
  * Can just grab this from running `deno tags > lib.deno.d.ts`
* It has it's own upgrade command
  * Just hope people dont touch it?/Let it not work?
  * Add a patch to remove or hide it?
* No versioning like existing node (and other language/interpreter) releases
  * Not of particular concern now but something to think about for future versions

---

**Tests of `deno` after install:**

```sh
deno --version
deno 1.0.0
v8 8.4.300
typescript 3.9.2
```

**Examples from help**

```
deno run https://deno.land/std/examples/welcome.ts
Welcome to Deno

deno eval "console.log(30933 + 404)"
31337
```

**Watch fs**

`watch.ts`:
```ts
const watcher = Deno.watchFs("./");
for await (const event of watcher) {
  console.log(">>>> event", event);
  // { kind: "create", paths: [ "/foo.txt" ] }
}
```

`deno run --allow-read watch.ts`

**Deno Packages/Programs**

`deno install --unstable --allow-read --allow-run -f https://deno.land/x/denon/denon.ts`

`~/.deno/bin/denon log.ts`

`log.ts`:
```ts
console.log("hello, world!")
```

### CODE REVIEW

**Comment by 06kellyjac** [comment]
Force-pushed `1.0.0-rc3`

**Comment by 06kellyjac** [comment]
I've finally got past `rusty_v8` while building deno with rust. Needed to include submodules, rebuilding now :crossed_fingers:

**Comment by 06kellyjac** [comment]
Force-pushed move to building from source, `rusty_v8` is still using the `.a` lib file from it's release, not building the whole of `v8` from source.

File: pkgs/development/web/deno/default.nix
Comment (prusnak) [inline pkgs/development/web/deno/default.nix:4]: Can we add `aarch64-unknown-linux-gnu` here?

File: pkgs/development/web/deno/default.nix
Comment (prusnak) [inline pkgs/development/web/deno/default.nix:57]: And `aarch64-linux` here?

**Review by prusnak — CHANGES_REQUESTED** [review]

File: pkgs/development/web/deno/default.nix
Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix:4]: Sure
I'm moving it over to a setup similar to firecracker ([link](https://github.com/NixOS/nixpkgs/blob/f83c5d3432d302125cd3f5ca2ee7f44934077fe0/pkgs/applications/virtualization/firecracker/default.nix)) because I noticed each of these libraries will have a different sha sum

**Comment by 06kellyjac** [comment]
Force-pushed completions for:
* `bash`
* `zsh`
* `fish`

Need to test them with `zsh` and `fish` but should work fine

File: pkgs/development/web/deno/default.nix
Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix]: The changes have been pushed but I've not tested them on Mac or an aarch64 arm machine yet. I'll resolve these conversations & track the testing progress above in the `"Built on platform(s)" section

**Comment by 06kellyjac** [comment]
Deno `1.0.0` release pr: https://github.com/denoland/deno/pull/5273

**Comment by 06kellyjac** [comment]
`aarch64-linux` & `x86_64-darwin` were the wrong way around

**Comment by 06kellyjac** [comment]
Currently building `1.0.0-rc3` on a raspberry pi4 aarm64

```
[~/projects/nixpkgs]$ uname -rmo
4.19.113-v8+ aarch64 GNU/Linux

[~/projects/nixpkgs]$ git status
On branch deno_release
Your branch is up to date with 'origin/deno_release'.

nothing to commit, working tree clean
[~/projects/nixpkgs]$ nix-env -f . -iA deno
...
Finished release [optimized] target(s) in 45m 19s
running tests
...
[in-progress]
```

**Comment by doronbehar** [comment]
Hey all.

I have absolutely no idea what deno is, but there's this someone who I've talked with at discourse who have made interesting progress regarding packaging deno while building from source:

https://discourse.nixos.org/t/deno-an-expression-in-need-of-love/6978

I think you'll be able to pick up something from there, while perhaps incorporating some of the advices I gave him.

The thing is, they seem to not want to open a GitHub account so they are not able to put their ideas directly here. I have no idea why but I don't judge. I'm happy to proxy his comments because they seem so nice :green_heart: .

**Comment by 06kellyjac** [comment]
Hi @doronbehar, thanks! I've had a bit of a scan through the discourse & I was going to read more indepth if I couldn't get `rusty_v8` to compile with it's static `.a` lib.

From my experience playing with building deno & from what I know about v8 is it's a pita to build from source. This build is already around 12mins long on my `i7-8850H (12) @ 4.3GHz` (need to time it) and 45mins long on a Raspberry Pi 4 `(4) @ 1.5GHz` ([see above](https://github.com/NixOS/nixpkgs/pull/87653#issuecomment-628164357).

As far as I can tell the node packages also don't build v8 from source:
[nodejs.nix common](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/web/nodejs/nodejs.nix)
[v14.nix version & sha for v0.14.2](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/web/nodejs/v14.nix)
[node-v14.2.0.tar.xz it downloads](https://nodejs.org/dist/v14.2.0/node-v14.2.0.tar.xz)
[hydra build of node v0.14.2](https://hydra.nixos.org/build/119103115)
[raw log from build of node v0.14.2](https://nix-cache.s3.amazonaws.com/log/0a3s7v2bnvg4gfjpzyysp93q0sw4kdb3-nodejs-14.2.0.drv)

`tools/make-v8.sh` doesn't seem to be called in the logs (other than it's shebang being replaced automatically)
And no mention of `V8_BUILD_OPTIONS`, `cd deps/v8`, `d8`, `v8_target_cpu=`, `out.gn` from that script.
(some of the eyecatching parts of those scripts)

[hydra build for v8](https://hydra.nixos.org/build/118954874)
[raw log from build of v8](https://nix-cache.s3.amazonaws.com/log/7qfwj2wvg4n5gq4iwfqfx74a3qf5gi56-v8-7.4.255.drv)

Also note the node build was "only" "1h 22m 21s" but just v8 on it's own is "2h 29m 29s"!!!

---

I could be wrong, there could be a load of tests in the v8 run that aren't needed for node etc. but this seems like more than enough depth for the build. Most package managers will just take the bin from the release and you're done. (Kinda like my [first iteration](https://github.com/NixOS/nixpkgs/commit/c869c1bf60b22b7b5cf34d6813ccd906d7e60d74))

I'll make a discourse account some time soon to make contacting me easier :slightly_smiling_face:

File: pkgs/development/web/deno/default.nix
Comment (andersk) [inline pkgs/development/web/deno/default.nix:44]: Nit: you mean `1>&2` (or `>&2`).

Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix:44]: Nice spot

Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix]: I've moved it to the following:

```sh
printf "%s\n" >&2 \
  "version mismatch between 'rusty_v8' in Cargo.lock and downloaded library:" \
  "  wanted: ${rustyV8Lib.meta.version}" \
  "  got:    $rusty_v8_ver"
```

My wanted and got were the wrong way around :sweat: 
Now pulls version directly from the lib (added `meta = { inherit version; };` to it's `fetchurl`)
The repeating `%s\n`s were superfluous.
The error msg and format are now more consistent with some of the other mismatch logs

```ini
# fetchgit example output:
hash mismatch in fixed-output derivation '/nix/store/mm332zdbgmmlvf14522af59nzidnry20-source':
  wanted: sha256:0qqsnbndb0p20w1rnglpv03b9l6iwm6vbwxzikd7fl2kvzxz5z3m
  got:    sha256:0k8mqy1hf9hkp60jhd0x4z814y36g51083b3r7prc69ih2523hd1

# previous output:
rusty_v8 version in Cargo.lock doesn't match downloaded release
Wanted: 0.4.2
Got:    0.4.1

# new output:
version mismatch between 'rusty_v8' in Cargo.lock and downloaded library:
  wanted: 0.4.1
  got:    0.4.2
```

**Review by andersk — COMMENTED** [review]
According to [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes), the commit message should be “deno: init at 1.0.0”. Otherwise looks good.

File: pkgs/development/web/deno/default.nix
Comment (prusnak) [inline pkgs/development/web/deno/default.nix:1]: Put these on separate lines - `nixpkgs-fmt` format will show you how.

File: pkgs/development/web/deno/default.nix
Comment (prusnak) [inline pkgs/development/web/deno/default.nix:36]: Maybe add a link to https://bugs.chromium.org/p/chromium/issues/detail?id=942720 which is the reason why Python 2.7 needs to be used. It will indicate a packager this can be changed once the issue is closed.

**Review by prusnak — APPROVED** [review]
Left couple of nits, but they are not blocking the merge.

**Comment by 06kellyjac** [comment]
> Left couple of nits, but they are not blocking the merge.

Please do add them as comments.

Either it'll merge as is now but I'll know what to fix for later, or I'll get it changed before it merges.
I'm happy either way.

File: pkgs/development/web/deno/default.nix
Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix:1]: I've ran `nixpkgs-fmt` but it didn't move them automatically. I've manually moved them & after a rerun they're still on their separate lines

File: pkgs/development/web/deno/default.nix
Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix:36]: Cleaned up the explanation & added the tracking link

**Review by prusnak — APPROVED** [review]

**Comment by 06kellyjac** [comment]
Not sure why that `run_v8_help` test failed...
The regex seems right, the output has d8 ...

```
[cli/tests/integration_tests.rs:2773] part = "\nSynopsis:\n"
[cli/tests/integration_tests.rs:2773] i = 1
[cli/tests/integration_tests.rs:2782] "exit false " = "exit false "
[cli/tests/integration_tests.rs:2782] i = 1
OUTPUT
Synopsis:
  shell [options] [--shell] [<file>...]
  d8 [options] [-e <string>] [--shell] [[--module] <file>...]
  -e        execute a string in V8
...
OUTPUT
EXPECTED
[WILDCARD]
Synopsis:
[WILDCARD]d8[WILDCARD]
EXPECTED
thread 'run_v8_help' panicked at 'pattern match failed', cli/tests/integration_tests.rs:2740:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by Gaelan** [comment]
Some small changes to make this build on Darwin:
```diff
diff --git a/pkgs/development/web/deno/default.nix b/pkgs/development/web/deno/default.nix
index 8cd2421800b..eb2669cd8b7 100644
--- a/pkgs/development/web/deno/default.nix
+++ b/pkgs/development/web/deno/default.nix
@@ -5,6 +5,9 @@
 , rustPlatform
 , python27
 , installShellFiles
+, Security
+, CoreServices
+, lib
 }:
 let
   pname = "deno";
@@ -48,6 +51,8 @@ rustPlatform.buildRustPackage rec {
     installShellFiles
   ];
 
+  buildInputs = lib.optionals stdenv.isDarwin [ Security CoreServices ];
+
   # The rusty_v8 package will try to download a `librusty_v8.a` release at build time to our read-only filesystem
   # To avoid this we pre-download the file and place it in the locations it will require it in advance
   preBuild = ''
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index db3491be140..b34d55f5f2f 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -2829,7 +2829,9 @@ in
 
   deer = callPackage ../shells/zsh/zsh-deer { };
 
-  deno = callPackage ../development/web/deno { };
+  deno = callPackage ../development/web/deno {
+    inherit (darwin.apple_sdk.frameworks) Security CoreServices;
+  };
 
   detox = callPackage ../tools/misc/detox { };
 
```

**Comment by prusnak** [comment]
No need to add `lib`, just use `stdenv.lib.optionals`

**Comment by 06kellyjac** [comment]
Thanks both. I'll get that patched in soon & tested on a mac

File: pkgs/development/web/deno/default.nix
Comment (marsam) [inline pkgs/development/web/deno/default.nix:82]:
```suggestion
    export HOME="$TMPDIR"
```

**Comment by marsam** [comment]
@GrahamcOfBorg build deno

File: pkgs/development/web/deno/default.nix
Comment (06kellyjac) [reply pkgs/development/web/deno/default.nix]: Ahh I got the original from here:
https://github.com/NixOS/nixpkgs/blob/b87096c7fbe75d6e43e8e5110347afe962c69963/pkgs/applications/networking/cluster/terraform/default.nix#L37
Updated to match https://nixos.org/nix/manual/#sec-common-env

Thanks :smiley:

**Comment by 06kellyjac** [comment]
Deno `1.0.1` launched an hour ago
Should I init at `1.0.1` or merge this & version bump after? 

---

I'll run a build on my machine either way

**Comment by Sorixelle** [comment]
Just a heads up - there's a `1.0.2` now :P Also, re the `.d.ts` - it's accessible just by running `deno types`, so there's no need to worry about packaging that.

**Comment by 06kellyjac** [comment]
@Sorixelle correct :+1:

Yeah I noticed that when reading the deno CI trying to add an aarch64 release for non-nix users

---
deno `1.0.2`
New `sha256`: `1sq6a4lph45s6sw4mjyv6p9y60cgfgq98jsyrp81kh32fs0xcqq7`
New `cargoSha256`: `0414s377iqa6rdlgc2d9rjhxbcsjzz0qcp9rva5ykxdw9xbsxv6y`
rustyv8 is still the same version so no changes there, built fine on x86_64 nixos

**Comment by marsam** [comment]
LGTM, thank you!

**Comment by 06kellyjac** [comment]
Thanks @marsam !

Now I'll add a pr for the 1.0.2 version bump :)

**Pull request merged**
